torchaudio/datasets/tedlium.py

```diff
@@ -0,0 +1,206 @@
+
+        Args:
+            fileid (str): File id to identify both text and audio files corresponding to the sample
+            line (int): Line identifier for the sample inside the text file
+            path (str): Dataset root path
+
+        Returns:
+            Tedlium_item: A namedTuple containing [waveform, sample_rate, transcript, talk_id, speaker_id, identifier]
+        """
+        transcript_path = os.path.join(path, "stm", fileid)
+        with open(transcript_path + ".stm") as f:
+            transcript = f.readlines()[line]
+            talk_id, _, speaker_id, start_time, end_time, identifier, transcript = transcript.split(" ", 6)
+
+        wave_path = os.path.join(path, "sph", fileid)
+        waveform, sample_rate = self._load_audio(wave_path + self._ext_audio, start_time=start_time, end_time=end_time)
+
+        return (waveform, sample_rate, transcript, talk_id, speaker_id, identifier)
+
+    def _load_audio(self, path: str, start_time: float, end_time: float, sample_rate: int = 16000) -> [Tensor, int]:
+        """Default load function used in TEDLIUM dataset, you can overwrite this function to customize functionality
+        and load individual sentences from a full ted audio talk file
+
+        Args:
+            path (str): Path to audio file
+            start_time (int, optional): Time in seconds where the sample sentence stars
+            end_time (int, optional): Time in seconds where the sample sentence finishes
+
+        Returns:
+            [Tensor, int]: Audio tensor representation and sample rate
+        """
+        start_time = int(float(start_time) * 16000)
+        end_time = int(float(end_time) * 16000)
+        if torchaudio.get_audio_backend() == "sox_io":
+            return torchaudio.load(path, frame_offset=start_time, num_frames=end_time - start_time)
+        return torchaudio.load(path)[:, start_time:end_time]
+
+    def __getitem__(self, n: int) -> Tuple[Tensor, int, str, int, int, int]:
+        """TEDLIUM dataset custom function overwritting default loadbehaviour.
+        Loads a TEDLIUM sample given a index N
+
+        Args:
+            n (int): Index of sample to be loaded
+
+        Returns:
+            Tedlium_item: A namedTuple containing [waveform, sample_rate, transcript, talk_id, speaker_id, identifier]
+        """
+        fileid, line = self._filelist[n]
+        return self._load_tedlium_item(fileid, line, self._path)
+
+    def __len__(self) -> int:
+        """DTEDLIUM dataset custom function overwritting len default behaviour.
+
+        Returns:
+            int: TEDLIUM dataset length
+        """
+        return len(self._filelist)
+
+    def get_phoneme_dict(self):
+        """Returns the phoneme dictionary of a TEDLIUM release
+
+        Returns:
+            dictionary: Phoneme dictionary for the current tedlium release
+        """
+        # Read phoneme dictionary
+        if not hasattr(self, "phoneme_dict"):
+            self.phoneme_dict = {}
```

Comment:
This design pattern might leave users questions like "Which one should I use, `phoneme_dict` or `get_phoneme_dict`?", "What are the differences?"

The following techniques are often used to improve this;
1. Use internal attribute name like `_phoneme_dict`.
    This does not prevent users to access the underlying object directly, but accessing an attribute prefixed with underscore clearly sends a message that user code is not doing the right thing. so it's okay.
2. Use `property` decorator.
    Lazy-initialization often fits well in `property` attribute.
    Something like this;

    ```python
    @property
    def phoneme_dict(self):
        if self._phoneme_dict is None:
            self._phoneme_dict = dict()
            # Fill the dictionary
        return self._phoneme_dict.copy()  # see the bellow for `copy`
    ```

Also instead of checking if it's initialized with `hasattr`, it's more straight forward to initialize the actual attribute object in constructor as `None` such as `self._phone_dict = None` and perform the check as `if self._phone_dict is None:`

Another concern is returning dictionary reference. The user code might accidentally modify the dictionary which can lead to a very subtle bug. Maybe it's better to return a copy of the dictionary. as `return self._phoneme_dict.copy()` (however this is still shallow copy so the value points to the original lists and user code can still modify the lists. maybe changing it to tuple might prevent such unintended modification, like `self.phoneme_dict[content[0]] = tuple(content[1:])`)

Reply:
Really good points, a commit with all the suggestions will follow :+1: